### PR TITLE
Lexer and Parser inplementation

### PR DESCRIPTION
**Pull request by lexdyel-mendez**
### Lexer and Parser 

Created the lexer and parser inside the lexer module. The same right now is very simple and doesn't contain any link to the intermediate code. It can run arithmetical and comparative expression. It is also able to assign variables. The grammar used so far is this one: 

Grammar:

- expr  : KEYWORD: var IDENTIFIER EQ expr
           | comp-expr ((KEYWORD:and|KEYWORD:or) comp-expr)*
- comp-expr  : not comp-expr
           | arith_expr ((EE|LT|GT|LTE|GTE) arith_expr)*
- arith-expr : term ((PLUS|MINUS) term)*
- term       : factor ((MUL|DIV) factor)*

- factor     : (PLUS|MINUS) factor
           | power
- power      : number (POW factor)*
- atom       : INT
           | FLOAT
           | IDENTIFIER
           | LPAREN expr RPAREN
           | if-expr
           | for-expr
           | while-expr
- if-expr    : KEYWORD:if expr KEYWORD:then expr
           | (KEYWORD:elif expr KEYWORD:then expr)*
           | (KEYWORD:else expr)?
- for-expr   : KEYWORD:for IDENTIFIER EQ expr KEYWORD:to expr
                  (KEYWORD:step expr)? KEYWORD:then expr
- while-expr : KEYWORD:while expr KEYWORD:then expr

### CODE REVIEW

File: src/lexer.py [hunk 1]
```diff
@@ -0,0 +1,1060 @@
...
+TT_EQ			= 'EQ'
+TT_LPAREN   	= 'LPAREN'
+TT_RPAREN   	= 'RPAREN'
+TT_EE			= 'EE'
+TT_NE			= 'NE'
+TT_LT			= 'LT'
+TT_GT			= 'GT'
+TT_LTE			= 'LTE'
+TT_GTE			= 'GTE'
+TT_EOF			= 'EOF'
```
Comment (TheParodicts) [inline src/lexer.py:122]: Not sure if this is an issue with Github, but, could these be lined up better?

File: src/lexer.py [hunk 2]
```diff
@@ -0,0 +1,1060 @@
...
+#######################################
+# CONSTANTS
+#######################################
+
+DIGITS = '0123456789'
+LETTERS = string.ascii_letters
+LETTERS_DIGITS = LETTERS + DIGITS
+
+#######################################
+# ERRORS
```
Comment (TheParodicts) [inline src/lexer.py:18]: If this isn't auto-generated, some comments/documentation on all the methods would greatly help to understand them.

File: src/lexer.py [hunk 3]
```diff
@@ -0,0 +1,1060 @@
...
+TT_RPAREN   	= 'RPAREN'
+TT_EE			= 'EE'
+TT_NE			= 'NE'
+TT_LT			= 'LT'
+TT_GT			= 'GT'
+TT_LTE			= 'LTE'
+TT_GTE			= 'GTE'
+TT_EOF			= 'EOF'
+
+KEYWORDS = [
```
Comment (TheParodicts) [inline src/lexer.py:125]: Descriptive comments would be helpful here.

File: src/lexer.py [hunk 4]
```diff
@@ -0,0 +1,1060 @@
...
+				tokens.append(self.make_equals())
+			elif self.current_char == '<':
+				tokens.append(self.make_less_than())
+			elif self.current_char == '>':
+				tokens.append(self.make_greater_than())
+			else:
+				pos_start = self.pos.copy()
+				char = self.current_char
+				self.advance()
+				return [], IllegalCharError(pos_start, self.pos, "'" + char + "'")
```
Comment (TheParodicts) [inline src/lexer.py:218]: A small comment under each of these would help identify what's happening much easier.

File: src/lexer.py [hunk 5]
```diff
@@ -0,0 +1,1060 @@
...
+				self.advance()
+			elif self.current_char == '(':
+				tokens.append(Token(TT_LPAREN, pos_start=self.pos))
+				self.advance()
+			elif self.current_char == ')':
+				tokens.append(Token(TT_RPAREN, pos_start=self.pos))
+				self.advance()
+			elif self.current_char == '!':
+				token, error = self.make_not_equals()
+				if error: return [], error
```
Comment (TheParodicts) [inline src/lexer.py:206]: Again, not sure if this is a gitHub error, but the indentation here seems off.

File: src/lexer.py [hunk 6]
```diff
@@ -0,0 +1,1060 @@
...
+		self.advance()
+
+	def advance(self):
+		self.pos.advance(self.current_char)
+		self.current_char = self.text[self.pos.idx] if self.pos.idx < len(self.text) else None
+
+	def make_tokens(self):
+		tokens = []
+
+		while self.current_char != None:
```
Comment (TheParodicts) [inline src/lexer.py]: the `!=` in python can be changed to `is not`, though that's a python convention. This works the same, but if you find it confusing to have all these symbols around, that could help.

File: src/lexer.py [hunk 7]
```diff
@@ -0,0 +1,1060 @@
...
+			if self.current_char in ' \t':
+				self.advance()
+			elif self.current_char in DIGITS:
+				tokens.append(self.make_number())
+			elif self.current_char in LETTERS:
+				tokens.append(self.make_identifier())
+			elif self.current_char == '+':
```
Comment (TheParodicts) [inline src/lexer.py:183]: Similar to my other comment, `==` can be changed to `is`

File: src/lexer.py [hunk 8]
```diff
@@ -0,0 +1,1060 @@
...
+				tokens.append(Token(TT_PLUS, pos_start=self.pos))
```
Comment (TheParodicts) [inline src/lexer.py:184]: I'm not sure what Token does, but that first parameter, `TT_PLUS` should also be preceded by it's method keyword, like you do with `pos_start=self.pos`

File: .vscode/settings.json [hunk 9]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (ChristianPerez34) [inline .vscode/settings.json:1]: Add a .gitignore for this

File: src/lexer.py [hunk 10]
```diff
@@ -0,0 +1,1060 @@
...
+			else:
+				pos_start = self.pos.copy()
+				char = self.current_char
+				self.advance()
+				return [], IllegalCharError(pos_start, self.pos, "'" + char + "'")
+
+		tokens.append(Token(TT_EOF, pos_start=self.pos))
+		return tokens, None
+
+	def make_number(self):
```
Comment (TheParodicts) [inline src/lexer.py:224]: Documentation. You can start it by typing `"""` underneath the def and hitting enter. That will autofill any parameters and returns, and all  you do is document what the method does.

File: src/lexer.py [hunk 11]
```diff
@@ -0,0 +1,1060 @@
...
+		num_str = ''
+		dot_count = 0
+		pos_start = self.pos.copy()
+
+		while self.current_char != None and self.current_char in DIGITS + '.':
```
Comment (TheParodicts) [inline src/lexer.py]: Here is a great place where replacing `!=` with `is not` would help readability.

Comment (TheParodicts) [inline src/lexer.py]: Additionally, I'm not sure what's happening in the second part of this "and", but some parenthesis would help readability. Also, a comment here would be suggested.

File: src/lexer.py [hunk 12]
```diff
@@ -0,0 +1,1060 @@
...
+
+	def as_string(self):
+		result  = f'{self.error_name}: {self.details}\n'
+		result += f'File {self.pos_start.fn}, line {self.pos_start.ln + 1}'
+		result += '\n\n' + string_with_arrows(self.pos_start.ftxt, self.pos_start, self.pos_end)
+		return result
+
+class IllegalCharError(Error):
+	def __init__(self, pos_start, pos_end, details):
+		super().__init__(pos_start, pos_end, 'Illegal Character', details)
```
Comment (ChristianPerez34) [inline src/lexer.py:38]:
```Python
super().__init__(pos_start, pos_end, 'Illegal Character')
```

File: src/lexer.py [hunk 13]
```diff
@@ -0,0 +1,1060 @@
...
+				dot_count += 1
+			num_str += self.current_char
+			self.advance()
+
+		if dot_count == 0:
+			return Token(TT_INT, int(num_str), pos_start, self.pos)
+		else:
+			return Token(TT_FLOAT, float(num_str), pos_start, self.pos)
+
+	def make_identifier(self):
```
Comment (TheParodicts) [inline src/lexer.py:244]: Documentation.

File: src/lexer.py [hunk 14]
```diff
@@ -0,0 +1,1060 @@
...
+		id_str = ''
+		pos_start = self.pos.copy()
+
+		while self.current_char != None and self.current_char in LETTERS_DIGITS + '_':
```
Comment (TheParodicts) [inline src/lexer.py:248]: Same comment as earlier for readability.

File: src/lexer.py [hunk 15]
```diff
@@ -0,0 +1,1060 @@
...
+		result += '\n\n' + string_with_arrows(self.pos_start.ftxt, self.pos_start, self.pos_end)
+		return result
+
+class IllegalCharError(Error):
+	def __init__(self, pos_start, pos_end, details):
+		super().__init__(pos_start, pos_end, 'Illegal Character', details)
+
+class ExpectedCharError(Error):
+	def __init__(self, pos_start, pos_end, details):
+		super().__init__(pos_start, pos_end, 'Expected Character', details)
```
Comment (ChristianPerez34) [inline src/lexer.py:42]:
```Python
super().__init__(pos_start, pos_end, 'Expected Character')
```

File: src/lexer.py [hunk 16]
```diff
@@ -0,0 +1,1060 @@
...
+
+	def make_identifier(self):
+		id_str = ''
+		pos_start = self.pos.copy()
+
+		while self.current_char != None and self.current_char in LETTERS_DIGITS + '_':
+			id_str += self.current_char
+			self.advance()
+
+		tok_type = TT_KEYWORD if id_str in KEYWORDS else TT_IDENTIFIER
```
Comment (TheParodicts) [inline src/lexer.py:252]: This is actually a really cool way of assigning values to a variable! I did not know about this!

File: src/lexer.py [hunk 17]
```diff
@@ -0,0 +1,1060 @@
...
+	def __init__(self, pos_start, pos_end, details):
+		super().__init__(pos_start, pos_end, 'Illegal Character', details)
+
+class ExpectedCharError(Error):
+	def __init__(self, pos_start, pos_end, details):
+		super().__init__(pos_start, pos_end, 'Expected Character', details)
+
+class InvalidSyntaxError(Error):
+	def __init__(self, pos_start, pos_end, details=''):
+		super().__init__(pos_start, pos_end, 'Invalid Syntax', details)
```
Comment (ChristianPerez34) [inline src/lexer.py:46]:
```Python
super().__init__(pos_start, pos_end, 'Invalid Syntax')
```

File: src/lexer.py [hunk 18]
```diff
@@ -0,0 +1,1060 @@
...
+		self.advance()
+
+		if self.current_char == '=':
+			self.advance()
+			tok_type = TT_GTE
+
+		return Token(tok_type, pos_start=pos_start, pos_end=self.pos)
+
+#######################################
+# NODES
```
Comment (TheParodicts) [inline src/lexer.py:304]: I'm not sure what a node is, so comments and documentation here would drastically help.

File: src/lexer.py [hunk 19]
```diff
@@ -0,0 +1,1060 @@
...
+	def __init__(self, pos_start, pos_end, details):
+		super().__init__(pos_start, pos_end, 'Expected Character', details)
+
+class InvalidSyntaxError(Error):
+	def __init__(self, pos_start, pos_end, details=''):
+		super().__init__(pos_start, pos_end, 'Invalid Syntax', details)
+
+class RTError(Error):
+	def __init__(self, pos_start, pos_end, details, context):
+		super().__init__(pos_start, pos_end, 'Runtime Error', details)
```
Comment (ChristianPerez34) [inline src/lexer.py:51]:
```Python
super().__init__(pos_start, pos_end, 'Runtime Error')
```

File: src/lexer.py [hunk 20]
```diff
@@ -0,0 +1,1060 @@
...
+# PARSE RESULT
+#######################################
+
+class ParseResult:
+	def __init__(self):
+		self.error = None
+		self.node = None
+		self.advance_count = 0
+
+	def register_advancement(self):
```
Comment (TheParodicts) [inline src/lexer.py:403]: Not sure if you wrote this or this is something premade, but why is this called "register_advancement"? It looks to advance the advance_count, so wouldn't "advance_register" be a better name?

File: src/lexer.py [hunk 21]
```diff
@@ -0,0 +1,1060 @@
...
+class InvalidSyntaxError(Error):
+	def __init__(self, pos_start, pos_end, details=''):
+		super().__init__(pos_start, pos_end, 'Invalid Syntax', details)
+
+class RTError(Error):
+	def __init__(self, pos_start, pos_end, details, context):
+		super().__init__(pos_start, pos_end, 'Runtime Error', details)
+		self.context = context
+
+	def as_string(self):
```
Comment (ChristianPerez34) [inline src/lexer.py:54]: Any reason why this method overrides super class ```as_string``` method?

File: src/lexer.py [hunk 22]
```diff
@@ -0,0 +1,1060 @@
...
+
+	def advance(self, ):
+		self.tok_idx += 1
+		if self.tok_idx < len(self.tokens):
+			self.current_tok = self.tokens[self.tok_idx]
+		return self.current_tok
+
+	def parse(self):
+		res = self.expr()
+		if not res.error and self.current_tok.type != TT_EOF:
```
Comment (TheParodicts) [inline src/lexer.py:438]: consistency with the `not` and the `!=`

File: src/lexer.py [hunk 23]
```diff
@@ -0,0 +1,1060 @@
...
+			return res.failure(InvalidSyntaxError(
+				self.current_tok.pos_start, self.current_tok.pos_end,
+				"Expected '+', '-', '*', '/', '^', '==', '!=', '<', '>', <=', '>=', 'and' or 'or'"
+			))
+		return res
+
+	###################################
+
+	def if_expr(self):
```
Comment (TheParodicts) [inline src/lexer.py:447]: Documentation

File: src/lexer.py [hunk 24]
```diff
@@ -0,0 +1,1060 @@
...
+		if self.current_tok.matches(TT_KEYWORD, 'else'):
+			res.register_advancement()
+			self.advance()
+
+			else_case = res.register(self.expr())
+			if res.error: return res
+
+		return res.success(IfNode(cases, else_case))
+
+	def for_expr(self):
```
Comment (TheParodicts) [inline src/lexer.py:507]: Documentation

File: src/lexer.py [hunk 25]
```diff
@@ -0,0 +1,1060 @@
...
+
+		res.register_advancement()
+		self.advance()
+
+		body = res.register(self.expr())
+		if res.error: return res
+
+		return res.success(ForNode(var_name, start_value, end_value, step_value, body))
+
+	def while_expr(self):
```
Comment (TheParodicts) [inline src/lexer.py:577]: Documentation

File: src/lexer.py [hunk 26]
```diff
@@ -0,0 +1,1060 @@
...
+
+		res.register_advancement()
+		self.advance()
+
+		body = res.register(self.expr())
+		if res.error: return res
+
+		return res.success(WhileNode(condition, body))
+
+	def atom(self):
```
Comment (TheParodicts) [inline src/lexer.py:607]: Documentation

File: src/lexer.py [hunk 27]
```diff
@@ -0,0 +1,1060 @@
...
+			while_expr = res.register(self.while_expr())
+			if res.error: return res
+			return res.success(while_expr)
+
+		return res.failure(InvalidSyntaxError(
+			tok.pos_start, tok.pos_end,
+			"Expected int, float, identifier, '+', '-', '('"
+		))
+
+	def power(self):
```
Comment (TheParodicts) [inline src/lexer.py:657]: documentation

File: src/lexer.py [hunk 28]
```diff
@@ -0,0 +1,1060 @@
...
+		return self.bin_op(self.atom, (TT_POW, ), self.factor)
+
+	def factor(self):
```
Comment (TheParodicts) [inline src/lexer.py:661]: Documentation

File: src/lexer.py [hunk 29]
```diff
@@ -0,0 +1,1060 @@
...
+		if tok.type in (TT_PLUS, TT_MINUS):
+			res.register_advancement()
+			self.advance()
+			factor = res.register(self.factor())
+			if res.error: return res
+			return res.success(UnaryOpNode(tok, factor))
+
+		return self.power()
+
+	def term(self):
```
Comment (TheParodicts) [inline src/lexer.py:674]: Documentation

File: src/lexer.py (hunk 1 above)
Comment (ChristianPerez34) [reply src/lexer.py:122]: For Pycharm users check the following boxes:
* Capture

For Atom users:
[Python-Auto PEP8](https://atom.io/packages/python-autopep8)

File: src/lexer.py [hunk 30]
```diff
@@ -0,0 +1,1060 @@
...
+		return self.bin_op(self.factor, (TT_MUL, TT_DIV))
+
+	def arith_expr(self):
```
Comment (TheParodicts) [inline src/lexer.py:678]: Documentation

File: src/lexer.py [hunk 31]
```diff
@@ -0,0 +1,1060 @@
...
+		return self.bin_op(self.term, (TT_PLUS, TT_MINUS))
+
+	def comp_expr(self):
```
Comment (TheParodicts) [inline src/lexer.py:682]: Documentation

File: src/lexer.py [hunk 32]
```diff
@@ -0,0 +1,1060 @@
...
+
+		if res.error:
+			return res.failure(InvalidSyntaxError(
+				self.current_tok.pos_start, self.current_tok.pos_end,
+				"Expected int, float, identifier, '+', '-', '(' or 'not'"
+			))
+
+		return res.success(node)
+
+	def expr(self):
```
Comment (TheParodicts) [inline src/lexer.py:723]: Documentation

File: src/lexer.py [hunk 33]
```diff
@@ -0,0 +1,1060 @@
...
+			return res.failure(InvalidSyntaxError(
+				self.current_tok.pos_start, self.current_tok.pos_end,
+				"Expected 'var', int, float, identifier, '+', '-', '(' or 'not'"
+			))
+
+		return res.success(node)
+
+	###################################
+
+	def bin_op(self, func_a, ops, func_b=None):
```
Comment (TheParodicts) [inline src/lexer.py:779]: Documentation

File: src/strings_with_arrows.py [hunk 34]
```diff
@@ -0,0 +1,26 @@
+def string_with_arrows(text, pos_start, pos_end):
```
Comment (TheParodicts) [inline src/strings_with_arrows.py:1]: What is `string_with_arrows`? I think this is the most puzzling bit of all the code. Documentation here would greatly help.

File: src/strings_with_arrows.py [hunk 35]
```diff
@@ -0,0 +1,26 @@
+def string_with_arrows(text, pos_start, pos_end):
+    result = ''
+
+    # Calculate indices
```
Comment (TheParodicts) [inline src/strings_with_arrows.py:4]: These comments, though humble, are great for helping understand what is happening.

File: src/lexer.py (hunk 6 above)
Comment (ChristianPerez34) [reply src/lexer.py]: When comparing values with ```None``` it is best to do what @TheParodicts mentioned. Therfore, if should look as follows

```Python
while self.current_char is not None:
```

You can also do this

```Python
while self.current_char:
```

**Review by TheParodicts — APPROVED** [review]
There is a severe lack of commenting and documentation that make this code very difficult to understand. That being said, if it properly works with the intermediate code, it gets my approval.

File: src/lexer.py (hunk 7 above)
Comment (ChristianPerez34) [reply src/lexer.py:183]: I believe it should stay as is.

**Review by ChristianPerez34 — CHANGES_REQUESTED** [review]

File: src/lexer.py (hunk 1 above)
Comment (lexdyel-mendez) [reply src/lexer.py:122]: Its a problem with GitHub, In my code they are aligned.

File: src/lexer.py (hunk 8 above)
Comment (lexdyel-mendez) [reply src/lexer.py:184]: does it have a huge effect?

**Comment by ChristianPerez34** [comment]
> 
> 
> There is a severe lack of commenting and documentation that make this code very difficult to understand. That being said, if it properly works with the intermediate code, it gets my approval.

**Review by ChristianPerez34 — APPROVED** [review]

**Review by alejoreyes96 — APPROVED** [review]

**Pull request merged**